src/bun.js/web_worker.zig

```diff
@@ -303,30 +276,29 @@ pub const WebWorker = struct {
         // always doing a first tick so we call CppTask without delay after dispatchOnline
         vm.tick();
 
-        {
-            while (true) {
-                while (vm.eventLoop().tasks.count > 0 or vm.active_tasks > 0 or vm.uws_event_loop.?.active > 0) {
-                    vm.tick();
-
-                    vm.eventLoop().autoTickActive();
-                }
-
-                if (!this.allowed_to_exit) {
-                    this.flushLogs();
-                    vm.eventLoop().tickPossiblyForever();
-                    continue;
+        while (true) {
+            while (vm.eventLoop().tasks.count > 0 or vm.active_tasks > 0 or vm.uws_event_loop.?.active > 0) {
+                this.eventloop_poll_ref.refConcurrently(this.parent);
+                vm.tick();
+                // When the worker is done, the VM is cleared, but we don't free
+                // the entire worker object, because this loop is still active and UAF will happen.
+                if (this.vm == null) {
+                    this.deinit();
+                    return;
                 }
 
-                vm.onBeforeExit();
-
-                if (!this.allowed_to_exit)
-                    continue;
+                vm.eventLoop().autoTickActive();
+            }
 
-                break;
+            if (this.vm == null) {
+                this.deinit();
+                return;
             }
 
             this.flushLogs();
-            this.onTerminate();
+            this.eventloop_poll_ref.unrefConcurrently(this.parent);
+            vm.eventLoop().tickPossiblyForever();
+            this.parent.eventLoop().wakeup();
```

Comment:
there is a race condition here where the parent event loop is unref'd

Reply:
for the case where the parent event loop is not the main thread

Reply:
oh i did not think about workers starting other workers